### PR TITLE
feat: implement event queue and consumer

### PR DESCRIPTION
**Pull request by adityathebe**

### CODE REVIEW

File: event_consumer.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+			time.Sleep(waitDurationOnFailure)
+		} else if count == 0 {
+			return
+		}
+	}
+}
+
+func (e *EventConsumer) Listen(ctx Context, pgNotify <-chan string) {
+	if err := e.Validate(); err != nil {
+		logger.Fatalf("error starting event consumer: %v", err)
```
Comment (moshloop) [inline event_consumer.go]: Return the error don't log

File: consumers/upstream.go [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+
+import (
+	"fmt"
+
+	"github.com/flanksource/commons/logger"
+	"github.com/flanksource/duty/upstream"
+	"github.com/flanksource/postq"
+)
+
+const EventPushQueueCreate = "push_queue.create"
```
Comment (moshloop) [inline consumers/upstream.go]: I think this should go into duty, I want to try and make postq a general library that we can publicise - which means we should not import duty or commons

File: event_consumer.go [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+		return
+	}
+
+	// Consume pending events
```
Comment (moshloop) [inline event_consumer.go]: Why do we do this - why not just let the background workers consume?

File: event_consumer.go [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+	}
+	return nil
+}
+
+// ConsumeEventsUntilEmpty consumes events in a loop until the event queue is empty.
+func (t *EventConsumer) ConsumeEventsUntilEmpty(ctx Context) {
+	for {
+		count, err := t.consumerFunc(ctx)
+		if err != nil {
+			logger.Errorf("error processing event, waiting %s to try again: %v", waitDurationOnFailure, err)
```
Comment (moshloop) [inline event_consumer.go]: I think we should abstract this away into a ErrorHandler e.g.
```go
EventConsumer {
          ErrorHandler: func(consumer Event Consumer, e error) bool {
logger.Infof("error")
time.Sleep(100);
return true; //retry 
}
}
```

File: postq.go [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+				ORDER BY priority DESC, created_at ASC
+				FOR UPDATE SKIP LOCKED
+				LIMIT @batchSize
+			)
+			RETURNING *
+		`
+
+	var events []Event
+	vals := map[string]any{
+		"maxAttempts": EventMaxAttempts,
```
Comment (moshloop) [inline postq.go]: I think maxAttempts, delay, exponential should all be parameters on EventConsumers with defaults

File: postq.go [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+	now := time.Now()
+	for i := range failedEvents {
+		e := failedEvents[i]
+		e.Attempts += 1
+		e.LastAttempt = &now
+	}
+
+	if err := failedEvents.Update(ctx, tx.Conn()); err != nil {
+		// TODO: More robust way to handle failed event insertion failures
+		log.Printf("error inserting into table 'event_queue': %v\n", err)
```
Comment (moshloop) [inline postq.go]: Shouldn't we just call the error handler ?

File: postq.go [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+	defer tx.Rollback(ctx)
+
+	events, err := fetchEvents(ctx, tx, t.WatchEvents, t.BatchSize, t.EventFetcherOpt)
+	if err != nil {
+		return 0, fmt.Errorf("error fetching events: %w", err)
+	}
+
+	failedEvents := t.Consumer(ctx, events)
+
+	// TODO: use db NOW()
```
Comment (moshloop) [inline postq.go]: I think we need to implement this before merging

File: postq.go (hunk 6 above)
Comment (adityathebe) [reply postq.go]: @moshloop The error handler added as mentioned in the previous discussion is for `PGConsumer` which is a separate independent entity from this async event queue handler _(playbook also uses it)._ 

We don't have error handler here and adding it doesn't make much sense as the error handling is supposed to be entirely handled by us _(increase attempts, setting errors ...)_

Comment (adityathebe) [reply postq.go]:
```
// TODO: More robust way to handle failed event insertion failures
log.Printf("error inserting into table 'event_queue': %v\n", err)
```

not sure how to handle this properly. Maybe we could retry a couple of times and if it still fails, we simply retry it again. I would say, any error in here is extremely rare.

**Pull request merged**